packages/dci-api/src/crypto/encrypt.ts

```diff
@@ -0,0 +1,78 @@
+import { FlattenedEncrypt, type FlattenedJWE, importJWK } from 'jose'
+import { type components } from '../registry-core-api'
+import { z } from 'zod'
+import { ValidationError, fromZodError } from 'zod-validation-error'
+
+function jweToEncryptedMessage(
+  jwe: FlattenedJWE
+): components['schemas']['EncryptedMessage'] {
+  if (jwe.header?.alg === undefined) {
+    throw new Error('Missing required header paramater "alg" after encrypting')
+  }
+  if (jwe.header.enc === undefined) {
+    throw new Error('Missing required header paramater "enc" after encrypting')
+  }
+  if (jwe.header.kid === undefined) {
+    throw new Error('Missing required header paramater "kid" after encrypting')
+  }
+  if (jwe.encrypted_key === undefined) {
+    throw new Error('encrypted_key is required after encrypting')
+  }
+  return {
+    ...jwe,
+    header: {
+      alg: jwe.header.alg,
+      enc: jwe.header.enc,
+      kid: jwe.header.kid,
+      ...jwe.header
+    },
+    encrypted_key: jwe.encrypted_key
+  }
+}
+
+const jwkSchema = z
+  .object({
+    kty: z.string(),
+    kid: z.string(),
+    use: z.enum(['sig', 'enc']),
+    alg: z.string()
+  })
+  .passthrough()
+
+const jwksSchema = z.object({
+  keys: z.array(jwkSchema)
+})
+
+export async function encryptPayload(
+  jwksUrl: string,
+  payload: components['schemas']['SearchResponse']
+): Promise<components['schemas']['EncryptedMessage']> {
+  const url = new URL(jwksUrl)
+  const response = await fetch(url)
```

Comment:
Would it make sense to cache this to not need to refetch the jwks every time? If JWKS gets updated, I think the stack gets restarted? Guess we could have something like:

```
function fetchJwks(url: string) {
  const url = new URL(jwksUrl)
  const response = await fetch(url)
  ...
}

const memoizedFetchJwks = memoize(fetchJwks) // https://lodash.com/docs/4.17.15#memoize
```

Reply:
yea ideally we could have some kind of a cache timeout like here: https://github.com/panva/jose/blob/358e864a0cccf1e0f9928a959f91f18f3f06a7de/docs/interfaces/jwks_remote.RemoteJWKSetOptions.md. Unfortunately, this is configured to be used with "signing" keys and not "encrytion" ones. So we would need to implement it ourselves I guess